tests/numerics/ad/test_secondary_operator.py

```diff
@@ -0,0 +1,300 @@
+
+    with pytest.raises(ValueError):
+        _ = sop.value_and_jacobian(eqsys)
+    with pytest.raises(ValueError):
+        _ = sop_pi.value_and_jacobian(eqsys)
+    with pytest.raises(ValueError):
+        _ = sop_pt.value_and_jacobian(eqsys)
+
+    ## Testing the setting of values using the local and global setter for iter values
+    for g in domains:
+        expr.progress_iterate_values_on_grid(np.zeros(g.num_cells), g)
+
+    # asserting the correct values are fetched by global getter and through parsing
+    # and shifting the values iteratively on domains
+    if on_intf:
+        assert np.all(np.zeros(nc) == expr.interface_values)
+        assert np.all(np.zeros(nc) == sop.value(eqsys))
+        expr.interface_values = np.ones(nc)
+    else:
+        assert np.all(np.zeros(nc) == expr.subdomain_values)
+        assert np.all(np.zeros(nc) == sop.value(eqsys))
+        expr.subdomain_values = np.ones(nc)
+
+    # Check parsing of previous iter and current op
+    assert np.all(sop_pi.value(eqsys) == np.zeros(nc))
+    assert np.all(sop.value(eqsys) == np.ones(nc))
+    # Still no data at previous time step
+    with pytest.raises(ValueError):
+        _ = sop_pt.value(eqsys)
+
+    # constructing diffs for current iter, with respective numbers of dependencies
+    if on_intf:
+        expr.interface_derivatives = np.array(diff_vals)
+    else:
+        expr.subdomain_derivatives = np.array(diff_vals)
+
+    # everything is set, the derivative values must have a certain structure
+    var_vals = [var.value_and_jacobian(eqsys) for var in vars]
+    sop_val = sop.value_and_jacobian(eqsys)
+
+    # Assert operator fetches the right values uppon evaluation
+    if on_intf:
+        assert np.allclose(sop_val.val, expr.interface_values)
+    else:
+        assert np.allclose(sop_val.val, expr.subdomain_values)
+
+    # The jacobians of the variables are identity blocks
+    # The Jacobian of the secondary operator must be the sum of respective identities
+    # with new data
+    jacs = []
+    for v, d in zip(var_vals, diff_vals):
+        jac_ = v.jac.copy()
+        jac_.data = d
+        jacs.append(jac_)
+
+    assert np.all(sop_val.jac.A == sum(jacs).A)
+
+    # progress values in time and check that only values are progressed, and that
+    # they are correct, i.e. current iter is set as previous time
+    expr.progress_values_in_time(domains)
+    if on_intf:
+        assert np.all(sop_pt.value(eqsys) == expr.interface_values)
+    else:
+        assert np.all(sop_pt.value(eqsys) == expr.subdomain_values)
+    # No derivatives progressed in time
+    for g in domains:
+        with pytest.raises(KeyError):
+            expr.fetch_data(sop_pt, g, True)
+
+    # previous iterate and time step have no derivative values, but the parsing should
+    # still work because the dependencies at previous iter and time are parsed as
+    # numpy arrays: The function of the SOP (AbstractFunction) should not call
+    # get_jacobian of the SOP
```

Comment:
This implies that `SecondaryExpression.fetch_data(..., True)` has behavior different from `SecondaryExpression.value_and_jacobian`. I understand why this is done, but is it wise to have two public methods for evaluation? An alternative would be to make `fetch_data` private (this would break with one standard for private methods, in that it will be called by other classes, but it would comply with the notion that private methods are part of the backend, not the API), another possibility is to overload `value` and `value_and_jacobian`. This brings us back to the question of where parsing should be located.

Reply:
`SecondaryExpression` has only `fetch_data` and it does only that, fetching data for specified time or iterate.
In `SecondaryExpression`, I would leave it public to give the user the opportunity to inspect the stored data.
Maybe make it private in the `SecondaryOPERATOR`? Because that class has `value_and_jacobian`, and only that one should be used.

Reply:
Agreed, let's make it private in SecondaryOperator (with an explanation why this is done).

Reply:
`SecondaryOperator._fetch_data` is now private. Updated documentation.